### PR TITLE
Feature/add more backend improvements

### PR DESCRIPTION
**Pull request by dawid-pstrak-rmtcfm-com**

### CODE REVIEW

File: backend/.env.example [hunk 1]
```diff
@@ -1,18 +1,19 @@
...
 PORT=8000
 APP_HOST=127.0.0.1
-APP_FRONTEND_URL=http://localhost:5000
+APP_FRONTEND_URL=http://localhost:3000
 APP_SECRET=8834dS537RD87faa43e9d33c5f5617174
 
 REDIS_HOST=127.0.0.1
 REDIS_PORT=6379
-REDIS_PASS=
+REDIS_PASSWORD=d329481234nsdf978324
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/.env.example]: for dev it can be without password

File: backend/controllers/Auth/LoginController.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+        const user = await this.userService.login(email, password);
+
+        if (!user) {
+            return res.sendStatus(this.httpStatusCodes.UNAUTHORIZED);
+        }
+
+        req.session.user = user;
+
+        return res.status(this.httpStatusCodes.OK).json(user);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Auth/LoginController.js]: res.send(user)

File: backend/controllers/Auth/LogoutController.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    async invoke(req, res, next) {
+        if (req.session) {
+            req.session.destroy(err => {
+                if (err) {
+                    next(err);
+                }
+            });
+        }
+
+        return res.sendStatus(this.httpStatusCodes.OK);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Auth/LogoutController.js]: NOT_FOUND

Comment (daniel-pupczyk-rmtcfm-com) [reply backend/controllers/Auth/LogoutController.js]: I menat NO_CONTENT :D

File: backend/controllers/Auth/RegisterController.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+class RegisterController {
+    constructor(userService, httpStatusCodes) {
+        this.userService = userService;
+        this.httpStatusCodes = httpStatusCodes;
+    }
+
+    async invoke(req, res) {
+        const { email, password, password_confirmation } = req.body;
+
+        if ((!email, !password || !password_confirmation || password !== password_confirmation)) {
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Auth/RegisterController.js]: are you sure there should be `comma`? plase add express-validator in some of next PR

File: backend/controllers/Auth/RegisterController.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+            return res.sendStatus(this.httpStatusCodes.BAD_REQUEST);
+        }
+
+        const user = await this.userService.register(email, password);
+
+        delete user.password;
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Auth/RegisterController.js]: refetch user again and then send him

File: backend/controllers/Auth/RegisterController.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+        req.session.user = user;
+
+        return res.status(this.httpStatusCodes.OK).json(user);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Auth/RegisterController.js]: res.send(user)

File: backend/controllers/Task/DestroyController.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+        this.httpStatusCodes = httpStatusCodes;
+    }
+
+    async invoke(req, res) {
+        const { id } = req.params;
+
+        const task = await this.taskService.findById(id);
+
+        if (!task) {
+            return res.sendStatus(404);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Task/DestroyController.js]: when not found in destroy then return 204

File: backend/controllers/Task/DestroyController.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        }
+
+        if (task.user_id !== req.session.user.id) {
+            return res.sendStatus(this.httpStatusCodes.FORBIDDEN);
+        }
+
+        await this.taskService.destroy(task._id, task._rev);
+
+        return res.sendStatus(this.httpStatusCodes.OK);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Task/DestroyController.js]: NO_CONTENT

File: backend/controllers/Task/DestroyController.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+        if (!task) {
+            return res.sendStatus(404);
+        }
+
+        if (task.user_id !== req.session.user.id) {
+            return res.sendStatus(this.httpStatusCodes.FORBIDDEN);
+        }
+
+        await this.taskService.destroy(task._id, task._rev);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Task/DestroyController.js]: just `await task.destroy()`

File: backend/controllers/Task/IndexController.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+class IndexController {
+    constructor(taskService, httpStatusCodes) {
+        this.taskService = taskService;
+        this.httpStatusCodes = httpStatusCodes;
+    }
+
+    async invoke(req, res) {
+        const tasks = await this.taskService.findAll({ user_id: request.session.user.id });
+
+        return res.status(this.httpStatusCodes.OK).json({ tasks });
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Task/IndexController.js]: res.send({ task })
also please fix responses and do not nest them, it should be `send(tasks)`

File: backend/controllers/Task/StoreController.js [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    async invoke(req, res) {
+        const { name } = req.body;
+
+        const task = await this.taskService.create({
+            name,
+            user_id: req.session.user.id,
+            created_at: Date.now()
+        });
+
+        return res.status(this.httpStatusCodes.CREATED).json(task);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Task/StoreController.js]: .status(201).send(task)

File: backend/controllers/Task/UpdateController.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+            return res.sendStatus(this.httpStatusCodes.FORBIDDEN);
+        }
+
+        await this.taskService.update({
+            ...task,
+            name,
+            completed
+        });
+
+        return res.sendStatus(this.httpStatusCodes.OK);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Task/UpdateController.js]: return updated task

File: backend/controllers/Task/UpdateController.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    }
+
+    async invoke(req, res) {
+        const { name, completed } = req.body;
+        const { id } = req.params;
+
+        const task = await this.taskService.findById(id);
+
+        if (!task) {
+            return res.sendStatus(404);
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Task/UpdateController.js]: use httpStatusCodes package

File: backend/.env.example (hunk 1 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/.env.example]: done

File: backend/controllers/Auth/LoginController.js (hunk 2 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Auth/LoginController.js]: done

File: backend/controllers/Auth/LogoutController.js (hunk 3 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Auth/LogoutController.js]: done

File: backend/controllers/Auth/RegisterController.js (hunk 4 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Auth/RegisterController.js]: yes, I'll. I just add di, split controllers and add smaller improvements in this PR. Code is copied for now

File: backend/controllers/Auth/RegisterController.js (hunk 6 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Auth/RegisterController.js]: done

File: backend/controllers/Task/DestroyController.js (hunk 7 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Task/DestroyController.js]: done

File: backend/controllers/Task/DestroyController.js (hunk 8 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Task/DestroyController.js]: done

File: backend/controllers/Task/DestroyController.js (hunk 9 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Task/DestroyController.js]: done

File: backend/controllers/Task/IndexController.js (hunk 10 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Task/IndexController.js]: done

File: backend/controllers/Task/StoreController.js (hunk 11 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Task/StoreController.js]: done

File: backend/controllers/Task/UpdateController.js (hunk 12 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Task/UpdateController.js]: done

File: backend/controllers/Task/UpdateController.js (hunk 13 above)
Comment (dawid-pstrak-rmtcfm-com) [reply backend/controllers/Task/UpdateController.js]: done

File: backend/controllers/Task/UpdateController.js [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+
+        if (!task) {
+            return res.sendStatus(this.httpStatusCodes.NOT_FOUND);
+        }
+
+        if (task.user_id !== req.session.user.id) {
+            return res.sendStatus(this.httpStatusCodes.FORBIDDEN);
+        }
+
+        await this.taskService.update({
```
Comment (daniel-pupczyk-rmtcfm-com) [inline backend/controllers/Task/UpdateController.js]: await task.update({ name, completed }

File: backend/controllers/Task/UpdateController.js [hunk 15]
```diff
@@ -0,0 +1,33 @@
+class UpdateController {
+    constructor(taskService, httpStatusCodes) {
+        this.taskService = taskService;
+        this.httpStatusCodes = httpStatusCodes;
+    }
+
+    async invoke(req, res) {
+        const { name, completed } = req.body;
```
Comment (ewelina-szczotka-rmtcfm-com) [inline backend/controllers/Task/UpdateController.js:8]: `const {body: {name, completed}, params: {id}} = req`

**Review by daniel-pupczyk-rmtcfm-com — APPROVED** [review]

File: backend/controllers/Auth/LoginController.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+        const user = await this.userService.login(email, password);
+
+        if (!user) {
+            return res.sendStatus(this.httpStatusCodes.UNAUTHORIZED);
+        }
+
+        req.session.user = user;
+
+        return res.send(user);
```
Comment (rafal-ksiazek-rmtcfm-com) [inline backend/controllers/Auth/LoginController.js:18]: without deleting password, you are sending encrypted password to the hacker, is this a deliberate procedure?

**Pull request merged**